### ISSUE TITLE
[data] Custom local shuffling batcher/sampler for Dataset.iter_*

### ISSUE THREAD

**Issue by wingkitlee0** [ray-project/ray#54197]
### Description

Allow customized "batcher" to be used in `iter_torch_batches`, `iter_batches` etc. Note: this seems to be related to the `Sampler` classes in pytorch (https://docs.pytorch.org/docs/stable/data.html#torch.utils.data.distributed.DistributedSampler).

See also https://github.com/ray-project/ray/issues/51392

Currently when `local_shuffle_buffer_size` is set, a `ShuffleBatcher` is used to shuffle a local buffer to avoid expensive global shuffle. However, the implementation is rather specific. It will be great if users can provide their implementations.

Also, there is already a base class `BatcherInterface` defined, which can be exposed as "DeveloperAPI". The usage should be similar to how user customizes `FileMetadataProvider` in some read APIs.

### Use case

It will be more flexible when users have some custom batching/epoch logic and want to have full control of their shuffling in the pipeline. For example, some models may be more sensitive to shuffling and it's not easy to keep increasing `local_shuffle_buffer_size`. In that case user can implement their own batcher..

**Comment by wingkitlee0** [comment]
My second attempt on this: https://github.com/ray-project/ray/pull/57196
which defines a simpler interface...

**Comment by repairman29** [comment]
Hey there! This looks like something I could help with. I'd be happy to take a look and see what we can do.

*(Found via [Echeo](https://echeo.io))*

**Comment by justinvyu** [comment]
Hey @wingkitlee0,

We recently introduced some new recommendations for scaling out the batching and collation steps in your data pipeline. @xinyuangui2 added [this user guide](https://docs.ray.io/en/latest/train/user-guides/scaling-collation-functions.html) with some examples.

High-level idea is to repartition your data into blocks of batch_size, and then to perform any final collation logic in map tasks with batch-size aligned blocks:

```python
read_parquet(...)
.map_batches(preprocess_fn)
.repartition(target_num_rows_per_block=BATCH_SIZE)  # distributed batching
.map_batches(collate_fn, batch_size=BATCH_SIZE)  # distributed collation
```

We are moving away from local_shuffle_buffer and iterator-side collation, since it runs into the scalability problems that you mentioned.

Let me know if this resolves some of the issues you mentioned, or if there are still some gaps.

**Comment by wingkitlee0** [comment]
@justinvyu So if I want to perform the same "local shuffle", e.g., with a large `local_shuffle_buffer_size`, do you recommend using you the pattern above and use the udf to do the random shuffle? 
I feel that they are not the same but maybe I am missing something (i.e., local shuffle in iter_batch currently has a buffer and perform random shuffle for each batch)

**Comment by richardliaw** [comment]
I would strongly suggest using the UDF for shuffle; we shouldn't open a custom implementation path here imo.
